clippy_utils/src/lib.rs

```diff
@@ -56,6 +56,81 @@ pub mod ty;
 pub mod usage;
 pub mod visitors;
 
+pub mod _internal {
+    use crate::source::snippet_opt;
+    use rustc_hir::{Expr, ExprKind};
+    use rustc_lint::LateContext;
+
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+    pub enum ExprPosition {
+        Closure,
+        Range,
+        AssignRhs,
+        AssignLhs,
+        OrLhs,
+        OrRhs,
+        AndLhs,
+        AndRhs,
+        EqLhs,
+        EqRhs,
+        BitOrLhs,
+        BitOrRhs,
+        BitXorLhs,
+        BitXorRhs,
+        BitAndLhs,
+        BitAndRhs,
+        ShiftLhs,
+        ShiftRhs,
+        AddLhs,
+        AddRhs,
+        MulLhs,
+        MulRhs,
+        Cast,
+        Prefix,
+        Postfix,
+    }
+
+    pub fn snip(cx: &LateContext<'_>, expr: &Expr<'_>, position: ExprPosition) -> String {
+        use ExprPosition::*;
+        let snip = match snippet_opt(cx, expr.span) {
+            Some(x) if has_enclosing_paren(&x) => return x,
+            Some(x) => x,
+            None => return "(..)".into(),
+        };
+        match expr.kind {
+            ExprKind::Unary(..) | ExprKind::AddrOf(..) | ExprKind::Box(..) if position > Prefix => {
+                format!("({})", snip)
+            },
+            ExprKind::Closure(..) | ExprKind::Ret(..) | ExprKind::Break(..) | ExprKind::Yield(..)
+                if position > Closure =>
```

Comment:
That would be equal to `position != Closure`, right?

Reply:
Yes.

Reply:
Then I'd prefer `!=`, because it's not susceptible to reordering.

Reply:
`ExprPostion` is already in precedence order and relies on it.

Reply:
Ok then.